python/ray/serve/_private/replica.py

```diff
@@ -727,10 +727,14 @@ async def perform_graceful_shutdown(self):
         # can skip the wait period.
         if self._user_callable_initialized:
             await self._drain_ongoing_requests()
-            await self._user_callable_wrapper.call_destructor()
 
         await self._metrics_manager.shutdown()
 
+        # We call the destructor last because the replica may not have
+        # initialized yet. The destructor may depend on instance variables
+        # that haven't been set yet, so it may raise an error.
```

Comment:
also, do we want to raise an error if the user's destructor raises an error? or just log and return? i don't think the exception needs to be propagated to the controller for any purpose

Reply:
I think we should just log and return. I left a GitHub comment about that, but it didn't get published :(

Reply:
We do currently pass the exception back to the controller ([code](https://github.com/ray-project/ray/blob/ab85dd292943a8b39b5d572f0e154e71c9d3d3d0/python/ray/serve/_private/deployment_state.py#L725-L732)).

Reply:
I think it's still useful to pass the exception back to the controller for logging purposes. If users aren't running an external logging system, then logging the exception replica-side means the log only exists while the worker node is still alive. If the node gets downscaled after the replica dies, the exception gets lost.

Reply:
I decided to just log and return. I don't like having an implicit requirement that the `__del__` step be the last part of the replica shutdown.